README_developers.md

```diff
@@ -91,128 +85,356 @@ $ # For a fix
+
+### Use K&R style
+
+Have the opening brace on the same line as the
+statement and the closing brace aligned with the statement.
+```c++
+// Yes: Use K&R braces
+while(a > 0) {
+  value += 1;
+}
+
+// No: Allman braces
+while(a > 0)
 {
   value += 1;
 }
 
-// Not preferred
-// please use braces
-if(some_condition)
+// No: GNU braces
+while(a > 0)
+  {
+    value += 1;
+  }
+```
+
+### Position of `else` keyword
+
+Keyword `else` should follow the if closing brace. 
+
+```c++
+// Yes: else keyword follows `if` closing brace
+if(a > 0) {
   value += 1;
+  b = a;
+} else {
+  value -= 1;
+}
+
+// No: else on the next line
+if(a > 0) {
+  value += 1;
+  b = a;
+}
+else {
+  value -= 1;
+}
+```
+### Use braces for every block
+
+For consistency and better code readability use braces
+for every code block in selection statements. Also,
+avoid condition and loop bodies in the same line of code
+```c++
+// Yes: Use braces for each block
+if(a > 0) {
+  value += 1;
+  b = a;
+} else {
+  value -= 1;
+}
+
+// No: Use braces only for blocks with more than one statement 
+if(a > 0) {
+  value += 1;
+  b = a;
+} else
+  value -= 1;
+
+// No: the condition and the body on the same line 
+if (a > 0) value += 1;
+else       value -= 1;
+```
 
-// Bad!
-if(some_condition) value += 1;
 
+### Do not use one true brace style
+
+For function and class definitions, have the opening brace
+on the next line aligned with the definition statement.
+```c++
+// Yes: opening brace on the next line for functions and classes declarations
+bool MyClass::my_method(int a)
+{
+
+}
+
+// No: opening brace on the same line
+bool MyClass::my_method(int a) {
+
+}
+``` 
+
+## RAJA specific style
+- Indent two spaces above surrounding scope
+- No spaces around template parameter
+- Line break before range segment
+- Use `RAJA_LAMBDA` define instead of specifying lambda function directly (e.g. `[=] __device__` )
+- Use `hiop_*` aliases over specific RAJA policies
+- Prefer `RAJA::Index_type` over `int` or anything else
+```c++
 // RAJA Lambdas
 RAJA::forall<hiop_raja_exec>(
   RAJA::RangeSegment(0, 10),
   RAJA_LAMBDA(RAJA::Index_type i)
   {
-    if(svec[i]==1.)
-    {
+    if(svec[i] == 1.) {
       local_data_dev[i] = c;
+    } else {
+      local_data_dev[i] = 0.;
     }
-    else
-    {
-      local_data_dev[i]=0.;
-    }
-  });
-
+  }
+);
 ```
 
 ## Naming Conventions
 
-- Classes
-  - Pascal case (upper and lower case, no underscores)
-- Data members
-  - Snake case
-    - lower case only
-    - use _ to separate
-  - Ending in _
-- Member methods
-  - Snake case
-    - lower case only
-    - use _ to separate
-- Avoid encoding type information in names
+### Classes
 
-For example:
+Use Pascal case for class names (upper and lower case, no underscores).
+When the name is composite of multiple words, each word starts with a capital
+letter. If one of the "words" is an acronym use same capitalization (first
+capital and other letters lowercase).
+```c++
+class MyHiopClass;
+```
```

Comment:
Does this mean we're going to move all classes to pascal case immediately (eg `hiopMatrixDense` -> `HiopMatrixDense`)? What's our timeline/plan to convert everything?

Reply:
That is a good point. I would pass the buck to @cnpetra here. My $0.02 would be to drop prefix `hiop` altogether from all HiOp class names because they are already in a `hiop` namespace.

Alternatively, we can just change the class names guideline so that each class name must begin with lowercase `hiop` like this:
```c++
class hiopMyHiopClass;
```

Reply:
> @cnpetra do we plan on rewriting all the legacy hiop code that uses old conventions? If so, how do we plan on going about it?
> 
> I wrote a quick clang tool to identify many of these changes and suggest replacements. When run on the following block:
> 
> ```c++
> class hiop_matrixDense {
>   int MyMemberVariable;
> };
> ```
> 
> it warns:
> 
> ```
> $ ./src/syntax/SyntaxCheckerStandalone --strip=hiop ../test/hiop.cpp
> ../test/hiop.cpp:2:1: warning: Found non-pascal case class/struct/union name. Please consider renaming.
> class hiop_matrixDense {
> ^     ~~~~~~~~~~~~~~~~
>       MatrixDense
> ../test/hiop.cpp:3:3: warning: Found non-snake case class/struct/union member variable name. Please consider renaming.
>   int MyMemberVariable;
>   ^   ~~~~~~~~~~~~~~~~
>       my_member_variable_
> 2 warnings generated.
> ```
> 
> It can auto-write the changes, but that seems a bit risky. Let me know what you think. cc @pelesh

this would be nice. It would be even nicer if we can run on the diff when one commits or when one does a PR.

While I think we should enforce the guidelines on all the new code, I am not sure if it will pay off to change all the code, given the risks and the fact that the code is kinda follows the guideline already...